Kernel/FileSystem/Plan9FileSystem.h

```diff
@@ -0,0 +1,352 @@
+ */
+
+#pragma once
+
+#include <AK/Atomic.h>
+#include <Kernel/FileSystem/FileBackedFileSystem.h>
+#include <Kernel/FileSystem/Inode.h>
+#include <Kernel/KBufferBuilder.h>
+
+namespace Kernel {
+
+class Plan9FSInode;
+
+class Plan9FS final : public FileBackedFS {
+    friend class Plan9FSInode;
+
+public:
+    virtual ~Plan9FS() override;
+    static NonnullRefPtr<Plan9FS> create(FileDescription&);
+
+    virtual bool initialize() override;
+
+    virtual bool supports_watchers() const override { return false; }
+
+    virtual NonnullRefPtr<Inode> root_inode() const override;
+
+    u16 allocate_tag() { return m_next_tag++; }
+    u32 allocate_fid() { return m_next_fid++; }
+
+    enum class ProtocolVersion {
+        v9P2000,
+        v9P2000u,
+        v9P2000L
+    };
+
+    struct qid {
+        u8 type;
+        u32 version;
+        u64 path;
+    };
+
+    class Message {
+    public:
+        static constexpr u16 NoTag = (u16)-1;
+
+        enum class Type : u8 {
+            // 9P2000.L
+            Tlerror = 6,
+            Rlerror = 7,
+            Tstatfs = 8,
+            Rstatfs = 9,
+
+            Tlopen = 12,
+            Rlopen = 13,
+            Tlcreate = 14,
+            Rlcreate = 15,
+            Tsymlink = 16,
+            Rsymlink = 17,
+            Tmknod = 18,
+            Rmknod = 19,
+            Trename = 20,
+            Rrename = 21,
+            Treadlink = 22,
+            Rreadlink = 23,
+            Tgetattr = 24,
+            Rgetattr = 25,
+            Tsetattr = 26,
+            Rsetattr = 27,
+
+            Txattrwalk = 30,
+            Rxattrwalk = 31,
+            Txattrcreate = 32,
+            Rxattrcreate = 33,
+
+            Treaddir = 40,
+            Rreaddir = 41,
+
+            Tfsync = 50,
+            Rfsync = 51,
+            Tlock = 52,
+            Rlock = 53,
+            Tgetlock = 54,
+            Rgetlock = 55,
+
+            Tlink = 70,
+            Rlink = 71,
+            Tmkdir = 72,
+            Rmkdir = 73,
+            Trenameat = 74,
+            Rrenameat = 75,
+            Tunlinkat = 76,
+            Runlinkat = 77,
+
+            // 9P2000
+            Tversion = 100,
+            Rversion = 101,
+            Tauth = 102,
+            Rauth = 103,
+            Tattach = 104,
+            Rattach = 105,
+            Terror = 106,
+            Rerror = 107,
+            Tflush = 108,
+            Rflush = 109,
+            Twalk = 110,
+            Rwalk = 111,
+            Topen = 112,
+            Ropen = 113,
+            Tcreate = 114,
+            Rcreate = 115,
+            Tread = 116,
+            Rread = 117,
+            Twrite = 118,
+            Rwrite = 119,
+            Tclunk = 120,
+            Rclunk = 121,
+            Tremove = 122,
+            Rremove = 123,
+            Tstat = 124,
+            Rstat = 125,
+            Twstat = 126,
+            Rwstat = 127
+        };
+
+        class Decoder {
```

Comment:
Does this really need to be in the header file? This header seems pretty bloated with implementation details..

Reply:
No, not really. And the header being way too bloated is one of the problems I see with this PR. I'll see if it's possible to move some of this into .cpp then.

Reply:
Moved all of `Message` (including `Message::Decoder`) into .cpp